File: invesdwin-scripting-parent/invesdwin-scripting-rust/invesdwin-scripting-rust-runtime-contract/src/test/java/de/invesdwin/scripting/rust/runtime/contract/callback/ParametersAndReturnsTestFloat.java
package de.invesdwin.scripting.rust.runtime.contract.callback;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.concurrent.NotThreadSafe;

import org.springframework.core.io.ClassPathResource;

import de.invesdwin.scripting.IScriptTaskEngine;
import de.invesdwin.scripting.IScriptTaskInputs;
import de.invesdwin.scripting.IScriptTaskResults;
import de.invesdwin.scripting.callback.IScriptTaskCallback;
import de.invesdwin.scripting.callback.ReflectiveScriptTaskCallback;
import de.invesdwin.scripting.rust.runtime.contract.AScriptTaskRust;
import de.invesdwin.scripting.rust.runtime.contract.IScriptTaskRunnerRust;
import de.invesdwin.util.assertions.Assertions;
import de.invesdwin.util.collections.Arrays;

@NotThreadSafe
public class ParametersAndReturnsTestFloat {

    private final IScriptTaskRunnerRust runner;

    public ParametersAndReturnsTestFloat(final IScriptTaskRunnerRust runner) {
        this.runner = runner;
    }

    public void testFloat() {
        new AScriptTaskRust<Void>() {

            @Override
            public IScriptTaskCallback getCallback() {
                final ParametersAndReturnsTestFloatCallback callback = new ParametersAndReturnsTestFloatCallback();
                return new ReflectiveScriptTaskCallback(callback);
            }

            @Override
            public void populateInputs(final IScriptTaskInputs inputs) {}

            @Override
            public void executeScript(final IScriptTaskEngine engine) {
                engine.eval(new ClassPathResource(ParametersAndReturnsTestFloat.class.getSimpleName() + ".rs",
                        ParametersAndReturnsTestFloat.class));
            }

            @Override
            public Void extractResults(final IScriptTaskResults results) {
                return null;
            }
        }.run(runner);
    }

    public static class ParametersAndReturnsTestFloatCallback {

        //putFloat
        private final float putFloat;

        //putFloatVector
        private final float[] putFloatVector;

        //putFloatVectorAsList
        private final List<Float> putFloatVectorAsList;

        //putFloatMatrix
        private final float[][] putFloatMatrix;

        //putFloatMatrixAsList
        private final List<List<Float>> putFloatMatrixAsList;

        public ParametersAndReturnsTestFloatCallback() {
            //putFloat
            this.putFloat = 123.123F;

            //putFloatVector
            this.putFloatVector = new float[3];
            for (int i = 0; i < putFloatVector.length; i++) {
                putFloatVector[i] = Float.parseFloat((i + 1) + "." + (i + 1));
            }

            //putFloatVectorAsList
            this.putFloatVectorAsList = Arrays.asList(Arrays.toObject(putFloatVector));

            //putFloatMatrix
            this.putFloatMatrix = new float[4][];
            for (int row = 0; row < putFloatMatrix.length; row++) {
                final float[] vector = new float[3];
                for (int col = 0; col < vector.length; col++) {
                    vector[col] = Float.parseFloat((row + 1) + "." + (col + 1));
                }
                putFloatMatrix[row] = vector;
            }

            //putFloatMatrixAsList
            this.putFloatMatrixAsList = new ArrayList<List<Float>>(putFloatMatrix.length);
            for (final float[] vector : putFloatMatrix) {
                putFloatMatrixAsList.add(Arrays.asList(Arrays.toObject(vector)));
            }
        }

        public float getFloat() {
            return putFloat;
        }

        public void setFloat(final float putFloat) {
            Assertions.assertThat(this.putFloat).isEqualTo(putFloat);
        }

        public float[] getFloatVector() {
            return putFloatVector;
        }

        public void setFloatVector(final float[] putFloatVector) {
            Assertions.assertThat(this.putFloatVector).isEqualTo(putFloatVector);
        }

        public List<Float> getFloatVectorAsList() {
            return putFloatVectorAsList;
        }

        public void setFloatVectorAsList(final List<Float> putFloatVectorAsList) {
            Assertions.assertThat(this.putFloatVectorAsList).isEqualTo(putFloatVectorAsList);
        }

        public float[][] getFloatMatrix() {
            return putFloatMatrix;
        }

        public void setFloatMatrix(final float[][] putFloatMatrix) {
            Assertions.assertThat(this.putFloatMatrix).isEqualTo(putFloatMatrix);
        }

        public List<List<Float>> getFloatMatrixAsList() {
            return putFloatMatrixAsList;
        }

        public void setFloatMatrixAsList(final List<List<Float>> putFloatMatrixAsList) {
            Assertions.assertThat(this.putFloatMatrixAsList).isEqualTo(putFloatMatrixAsList);
        }

    }

}

File: invesdwin-scripting-parent/invesdwin-scripting-runtime-lua/src/test/java/de/invesdwin/scripting/runtime/lua/tests/callback/ParametersAndReturnsTestFloat.lua
arraysToString = java.method(java.import('de.invesdwin.util.collections.Arrays'),'toString','java.lang.Object')

print("getFloat")
if(getFloat ~= nil) then
	error("getFloat already defined!")
end
getFloat = callback("getFloat")
getFloatType = type(getFloat)
print(getFloatType)
print(getFloat)
if(getFloatType ~= "number") then
	error("getFloat not Float!")
end
callback("setFloat", getFloat)

print("getFloatVector")
if(getFloatVector ~= nil) then
	error("getFloatVector already defined!")
end
getFloatVector = callback("getFloatVector")
getFloatVectorType = type(getFloatVector[1])
print(getFloatVectorType)
print(arraysToString(getFloatVector))
if(getFloatVectorType ~= "number") then
	error("getFloatVector not Float!")
end
callback("setFloatVector", getFloatVector)

print("getFloatVectorAsList")
if(getFloatVectorAsList ~= nil) then
	error("getFloatVectorAsList already defined!")
end
getFloatVectorAsList = callback("getFloatVectorAsList")
getFloatVectorAsListType = type(getFloatVectorAsList[1])
print(getFloatVectorAsListType)
print(arraysToString(getFloatVectorAsList))
if(getFloatVectorAsListType ~= "number") then
	error("getFloatVectorAsList not Float!")
end
callback("setFloatVectorAsList", getFloatVectorAsList)

print("getFloatMatrix")
if(getFloatMatrix ~= nil) then
	error("getFloatMatrix already defined!")
end
getFloatMatrix = callback("getFloatMatrix")
getFloatMatrixType = type(getFloatMatrix[1][1])
print(getFloatMatrixType)
print(arraysToString(getFloatMatrix))
if(getFloatMatrixType ~= "number") then
	error("getFloatMatrix not Float!")
end
callback("setFloatMatrix", getFloatMatrix)

print("getFloatMatrixAsList")
if(getFloatMatrixAsList ~= nil) then
	error("getFloatMatrixAsList already defined!")
end
getFloatMatrixAsList = callback("getFloatMatrixAsList")
getFloatMatrixAsListType = type(getFloatMatrixAsList[1][1])
print(getFloatMatrixAsListType)
print(arraysToString(getFloatMatrixAsList))
if(getFloatMatrixAsListType ~= "number") then
	error("getFloatMatrixAsList not Float!")
end
callback("setFloatMatrixAsList", getFloatMatrixAsList)
